### DISCUSSION TITLE
Log resolver exceptions to Sentry

### DISCUSSION THREAD

**Discussion by nilansaha** [mirumee/ariadne#685] [Q&A]
I recently got to know that it is a part of GraphQL Spec to eat up any exceptions from inside resolvers and return them as part of errors key in GraphQL result. Is there any way to raise the exception from inside the resolver anyway and log it to Sentry ?

**Comment by rafalp** [comment]
Have you tested if those exceptions aren't logged to Sentry? Sentry SDK plugs itself into Python's logging stack and it should log exceptions without extra work on Ariadne's side.

**Reply by nilansaha** [reply]
I am looking into it but I have another confusion as well. It seems errors are logged twice like this.
```
Cannot query field 'idd' on type 'Todo'. Did you mean 'id'?

GraphQL request:4:7
3 |     todo {
4 |       idd
  |       ^
5 |     }
graphql.error.graphql_error.GraphQLError: Cannot query field 'idd' on type 'Todo'. Did you mean 'id'?

GraphQL request:4:7
3 |     todo {
4 |       idd
  |       ^
5 |     }
```
Any idea why ? Or how I can stop the second one ?

**Reply by rafalp** [reply]
You will need to implement custom filter for Sentry's logger to get rid of this. Sentry SDK has `before_send` hook that allows you to inject custom function that takes exception and is able to convert it into different form:

```python
def prepare_exc_for_sentry(event, hint):
    if "exc_info" in hint:
        exc_type, exc_value, tb = hint["exc_info"]
        if exc_type != GraphQLError:
            return event

        top_exception = event["exception"]["values"][-1]
        if "GraphQL request" in top_exception["value"]:
            # Remove query fragment from exception message
            query_offset = top_exception["value"].find("GraphQL request")
            if query_offset > 5:
                top_exception["value"] = top_exception["value"][:query_offset].strip()

    return event
```
